# Classificador base - Majority learner
O classificador Majority Learner é um tipo simples de classificador utilizado em aprendizado de máquina. Ele funciona de forma bastante intuitiva: ele simplesmente atribui a classe mais frequente (a classe majoritária) a todas as instâncias do conjunto de dados de entrada.

A ideia básica por trás do Majority Learner é que, se uma classe é mais comum do que as outras no conjunto de dados, então atribuir essa classe a todas as instâncias como uma previsão provavelmente terá uma precisão razoável.

**Pontos Fortes do Majority Learner:**

1. **Simplicidade:** É extremamente simples de entender e implementar, exigindo pouca complexidade algorítmica.
  
2. **Rápido:** Como não envolve cálculos complexos, é muito rápido para treinar e fazer previsões.

3. **Baseline de Referência:** Serve como um ponto de referência básico para comparar o desempenho de outros classificadores mais complexos.

4. **Adequado para Classes Desbalanceadas:** Pode funcionar bem em conjuntos de dados com classes desbalanceadas, onde uma classe é significativamente mais frequente do que outras.

**Pontos Fracos do Majority Learner:**

1. **Limitado a Dados Balanceados:** Não é adequado para conjuntos de dados onde as classes têm distribuições muito desiguais, pois tende a favorecer a classe majoritária.

2. **Baixa Precisão:** Em conjuntos de dados complexos ou com distribuição desigual das classes, pode ter uma precisão muito baixa, pois atribui a mesma classe à maioria das instâncias.

3. **Não Considera Informações do Recurso:** Não leva em conta as características individuais das instâncias, o que pode resultar em previsões imprecisas em conjuntos de dados complexos.

4. **Não Adaptativo:** Não aprende com os dados nem ajusta seu comportamento com base nas características do conjunto de dados; portanto, não é adequado para problemas que requerem generalização sofisticada.

## Base credit data - 85.85%

In [ ]:
!pip install orange3

In [ ]:
import Orange

In [ ]:
# carregar base de dados com orange.data.Table
base_credit = Orange.data.Table('/content/drive/MyDrive/ML e Data Sciece com python/dataset/credit_risc/credit_data_regras.csv')

In [ ]:
# Consultar o domínio (esquema) da base de dados
base_credit.domain

[income, age, loan | default]

a notação [income, age, loan | default] indica que temos três atributos (renda, idade e empréstimo) e a variável alvo (padrão) nesta base de dados.

In [ ]:
# Importar o classificador Majority Learner do pacote Orange para classificação
import Orange.classification

# Criar uma instância do classificador Majority Learner
majority = Orange.classification.MajorityLearner()


In [ ]:
# Importar o módulo de avaliação e teste do pacote Orange
import Orange.evaluation.testing

# Realizar teste na base de dados utilizando o classificador Majority Learner
# A base de dados de teste e treinamento são as mesmas (base_credit)
# O classificador utilizado é o Majority Learner
previsoes = Orange.evaluation.testing.TestOnTestData(base_credit, base_credit, [majority])


**IMPORTANTE: NÃO EXISTE DE FATO UM TREINAMENTO COM ESSE ALGORITMO**

In [ ]:
# Importar a métrica de acurácia (CA - Classification Accuracy) do módulo de avaliação do Orange
import Orange.evaluation

# Calcular a acurácia utilizando as previsões geradas pelo teste
Orange.evaluation.CA(previsoes)


array([0.8585])

array([0.8585]) representa a acurácia do modelo de classificação

In [ ]:
# Iterar sobre cada registro na base de dados base_credit
for registro in base_credit:
    # Obter e imprimir a classe de cada registro usando o método get_class()
    print(registro.get_class())


0
0
0
0
1
0
0
1
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
1
0
0
0
1
0
0
0
0
0
0
0
0
0
0
0
0
1
0
0
0
0
0
0
0
0
0
0
1
0
0
0
1
1
0
0
0
0
0
0
0
0
0
0
0
1
1
1
0
0
0
0
0
0
0
0
0
0
0
1
1
0
0
0
1
0
0
1
0
0
0
0
0
0
0
0
0
0
0
0
0
0
1
0
0
0
0
0
0
0
0
0
0
1
0
0
0
0
1
0
1
1
0
1
0
0
1
0
0
0
0
0
0
0
0
0
0
0
1
0
0
0
0
0
0
1
0
1
0
0
0
0
0
0
0
0
0
0
1
0
1
1
0
0
0
1
1
0
0
0
0
1
0
0
0
0
1
0
0
0
0
0
1
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
1
0
0
1
1
0
0
0
1
0
1
0
0
0
0
0
0
0
0
1
0
0
0
1
0
0
0
0
0
0
0
0
0
1
0
0
0
0
0
0
0
0
0
0
0
0
1
0
1
0
1
0
1
1
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
1
0
0
0
1
1
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
1
0
1
0
0
0
0
0
0
0
0
0
0
0
0
0
0
1
0
0
1
0
0
0
1
0
0
0
1
0
0
0
0
0
0
0
1
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
1
0
0
0
0
0
0
0
0
0
0
1
0
0
0
0
0
1
0
0
1
0
0
0
0
0
0
1
0
0
1
0
0
0
0
0
1
0
0
0
0
0
1
0
1
0
0
0
1
0
1
0
0
0
0
1
0
0
0
0
0
0
0
0
0
0
0
1
0
1
0
0
1
1
0
0
0
0
0
1
0
0
0
0
0
0
0
1
0
0
0
0
0
0
0
1
0
1
0
0
0
0
0
1
0
0
0
1
0
0
1
0
0
0
0


In [ ]:
from collections import Counter

# Utiliza a classe Counter para contar a frequência de cada classe na variável alvo da base de dados
Counter(str(registro.get_class()) for registro in base_credit)

Counter({'0': 1717, '1': 283})

O retorno Counter({'0': 1717, '1': 283}) representa a contagem de ocorrências de cada classe na variável alvo da base de dados.

* Classe '0': Existem 1717 instâncias na base de dados que têm a classe '0'.

* Classe '1': Existem 283 instâncias na base de dados que têm a classe '1'.

In [ ]:
# Cálculo da proporção da classe '0' em relação ao total de instâncias na base de dados
proporcao_classe_0 = 1717 / 2000
proporcao_classe_0

0.8585

0.8585 representa a proporção da classe '0' em relação ao total de instâncias na base de dados

## Base census - 75.91%

In [ ]:
base_census = Orange.data.Table('/content/drive/MyDrive/ML e Data Sciece com python/dataset/census/census_regras.csv')

In [ ]:
base_census.domain

[age, workclass, final-weight, education, education-num, marital-status, occupation, relationship, race, sex, capital-gain, capital-loos, hour-per-week, native-country | income]

In [ ]:
majority = Orange.classification.MajorityLearner()
previsoes = Orange.evaluation.testing.TestOnTestData(base_census, base_census, [majority])
Orange.evaluation.CA(previsoes)

array([0.75919044])

In [ ]:
Counter(str(registro.get_class()) for registro in base_census)

Counter({'<=50K': 24720, '>50K': 7841})

In [ ]:
24720 / (24720 + 7841)

0.7591904425539756